RFC.md

```diff
@@ -0,0 +1,466 @@
+
+**TLDR**
+- Multimodal models are primarily DeepFusion and EarlyFusion architectures
+- Modules/utils for fusing and un-fusing MM models can allow for fine-tuning a variety of MM architectures in a uniform way.
+
+## Proposal
+
+**High Level List:**
+- Introduce a modules/model_fusion folder for handling fused models
+- fusion_models introduces two top level wrapper modules that combine a single encoder and decoder in two different ways. This allows the two models to be used as one model.
+- fusion_layer and fusion_embed are fusion modules (similar to PeFT modules) that introduce extra trainable parameters to the pre-trained fused models
+- fusion_utils is based on peft_utils and provides functions for accessing and modifying fusion parameters
+
+**Folder Structure for new modules/functions**
+```
+torchtune/
+	modules/
+		model_fusion/
+			fusion_models.py
+				DeepFusionModel
+				EarlyFusionModel
+			fusion_layer.py
+			fusion_embed.py
+			fusion_utils.py
+				register_fusion_module
+				get_fusion_params
+				set_trainable_params
+```
+
+**Pseudo code demonstrating usage of fused models**
+```python
+def flamingo_builder():
+	return DeepFusionModel(
+		encoder=CLIP(...),
+		decoder=Llama3(...))
+
+>>> model = flamingo_builder()
+
+# Load full checkpoints
+>>> model.load_state_dict(...)
+
+# Or load pretrained individual models
+>>> model.encoder.load_state_dict(...)
+>>> model.decoder.load_state_dict(...)
+
+# set only the fusion parameters to be trainable
+>>> fusion_params = get_fusion_params(model)
+>>> set_trainable_params(model, fusion_params)
+
+# Do inference with the model the same as a text model
+# Fusion models have the same signatures as TransformerDeocder
+>>> data = flamingo_transform(...) # Message formated text + images
+>>> model(**data)
+tensor([-1.0885, ..., .0231], grad_fn=<ViewBackward0>)
+```
+
+> [!Note]
+> This RFC proposes a complete design for supporting fusion models, but not all of the design is required to be built to support any one given fusion style model. This RFC gives us an opportunity to take a holistic view over the space, but it is recommended that each module/component is built as need.
+### Fusion Models
+DeepFusion requires that Decoder is already modified to have any additional layers for accepting encoder inputs (these should be wrapped with FusionLayer to be able to track these additional parameters and work with checkpoints). For inference all of the cache can be managed by the decoder, so these methods are simply passed through.
+
+```python
+class DeepFusionModel(nn.Module):
+	def __init__(self,
+		decoder: TransformerDecoder,
+		encoder: nn.Module,
+		...
+	):
+
+	def setup_caches(self, batch_size: int, dtype: torch.dtype) -> None:
```

Comment:
OK I'm starting to hate how much we have to bubble this up everywhere. I know I literally just reviewed the PR like two days ago, but can you remind me if there's anything stopping us from defining setup_caches as a function taking in a TransformerDecoder (or something) as an arg? Then we could just call setup_caches(model.decoder, batch_size, dtype) in our generate code or wherever

Reply:
Also not a fan of how ubiquitous these methods are. You need them defined for layer objects, as layers can have custom logic. But the TransformerDecoder (and it's wrappers) wouldn't need to have them, we could add a function like you said to generation utils. I'd argue that could be it's own PR outside of the scope of this RFC.

Reply:
Agree on a separate RFC, I think there are a few diff considerations